File: src/test/java/leets/leenk/domain/feed/domain/service/FeedGetServiceTest.java
package leets.leenk.domain.feed.domain.service;

import leets.leenk.domain.feed.application.exception.FeedNotFoundException;
import leets.leenk.domain.feed.domain.entity.Feed;
import leets.leenk.domain.feed.domain.repository.FeedRepository;
import leets.leenk.domain.feed.domain.service.dto.FeedNavigationResult;
import leets.leenk.domain.feed.test.FeedTestFixture;
import leets.leenk.domain.feed.test.UserTestFixture;
import leets.leenk.domain.user.domain.entity.User;
import leets.leenk.domain.user.domain.entity.UserBlock;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.data.domain.SliceImpl;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static leets.leenk.domain.feed.test.FeedTestFixture.createFeedWithCreateDate;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;

@ExtendWith(MockitoExtension.class)
public class FeedGetServiceTest {
    @Mock
    private FeedRepository feedRepository;

    @InjectMocks
    private FeedGetService feedGetService;

    @Test
    @DisplayName("findById 테스트")
    void findById1() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);

        given(feedRepository.findByDeletedAtIsNullAndId(1L))
                .willReturn(Optional.of(feed));

        // when
        Feed result = feedGetService.findById(1L);

        // then
        assertThat(result).isSameAs(feed);
        then(feedRepository).should().findByDeletedAtIsNullAndId(1L);
    }

    @Test
    @DisplayName("findById 테스트 - 예외")
    void findById2() {
        // given
        given(feedRepository.findByDeletedAtIsNullAndId(99L))
                .willReturn(Optional.empty());

        // when & then
        assertThatThrownBy(() -> feedGetService.findById(99L))
                .isInstanceOf(FeedNotFoundException.class);
        then(feedRepository).should().findByDeletedAtIsNullAndId(99L);
    }

    @Test
    @DisplayName("findAll 테스트")
    void findAll() {
        // given
        User blocker = UserTestFixture.createUser(1L, "blocker");
        User blocked1 = UserTestFixture.createUser(2L, "blocked1");
        User blocked2 = UserTestFixture.createUser(3L, "blocked2");
        List<UserBlock> blockedUsers = List.of(
                UserBlock.builder().blocker(blocker).blocked(blocked1).build(),
                UserBlock.builder().blocker(blocker).blocked(blocked2).build()
        );

        Pageable pageable = PageRequest.of(0, 10);
        Slice<Feed> slice = new SliceImpl<>(List.of());

        given(feedRepository.findAllByDeletedAtIsNullWithUser(pageable, List.of(2L, 3L)))
                .willReturn(slice);

        // when
        Slice<Feed> result = feedGetService.findAll(pageable, blockedUsers);

        // then
        assertThat(result).isSameAs(slice);
        then(feedRepository).should().findAllByDeletedAtIsNullWithUser(pageable, List.of(2L, 3L));
    }

    @Test
    @DisplayName("findAllByUser 테스트")
    void findAllByUser() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Pageable pageable = PageRequest.of(0, 5);
        Slice<Feed> slice = new SliceImpl<>(List.of());

        given(feedRepository.findAllByUserAndDeletedAtIsNull(user, pageable))
                .willReturn(slice);

        // when
        Slice<Feed> result = feedGetService.findAllByUser(user, pageable);

        // then
        assertThat(result).isSameAs(slice);
        then(feedRepository).should().findAllByUserAndDeletedAtIsNull(user, pageable);
    }

    @Test
    @DisplayName("findPrevFeedsWithHasMore 테스트")
    void findPrevFeedsWithHasMore() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        LocalDateTime base = LocalDateTime.of(2025, 12, 22, 16, 0);
        Feed current = createFeedWithCreateDate(100L, user, base);
        Feed first = createFeedWithCreateDate(1L, user, base.plusMinutes(1));
        Feed second = createFeedWithCreateDate(2L, user, base.plusMinutes(2));
        Feed third = createFeedWithCreateDate(3L, user, base.plusMinutes(3));

        given(feedRepository.findPrevFeeds(eq(base), isNull(), eq(PageRequest.of(0, 3))))
                .willReturn(new ArrayList<>(List.of(first, second, third)));

        // when
        FeedNavigationResult result = feedGetService.findPrevFeedsWithHasMore(current, List.of(), 2);

        // then
        assertThat(result.hasMore()).isTrue();
        assertThat(result.feeds()).containsExactly(second, first);
        then(feedRepository).should().findPrevFeeds(eq(base), isNull(), eq(PageRequest.of(0, 3)));
    }

    @Test
    @DisplayName("findNextFeedsWithHasMore 테스트")
    void findNextFeedsWithHasMore() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        User blocked = UserTestFixture.createUser(2L, "blocked");
        UserBlock block = UserBlock.builder().blocker(user).blocked(blocked).build();
        LocalDateTime base = LocalDateTime.of(2025, 12, 22, 16, 0);
        Feed current = createFeedWithCreateDate(100L, user, base);
        Feed first = createFeedWithCreateDate(1L, user, base.minusMinutes(1));
        Feed second = createFeedWithCreateDate(2L, user, base.minusMinutes(2));
        Feed third = createFeedWithCreateDate(3L, user, base.minusMinutes(3));

        given(feedRepository.findNextFeeds(eq(base), eq(List.of(2L)), eq(PageRequest.of(0, 3))))
                .willReturn(new ArrayList<>(List.of(first, second, third)));

        // when
        FeedNavigationResult result = feedGetService.findNextFeedsWithHasMore(current, List.of(block), 2);

        // then
        assertThat(result.hasMore()).isTrue();
        assertThat(result.feeds()).containsExactly(first, second);
        then(feedRepository).should().findNextFeeds(eq(base), eq(List.of(2L)), eq(PageRequest.of(0, 3)));
    }
}

File: src/test/kotlin/leets/leenk/domain/feed/domain/service/FeedGetServiceTest.kt
package leets.leenk.domain.feed.domain.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import leets.leenk.domain.feed.application.exception.FeedNotFoundException
import leets.leenk.domain.feed.domain.entity.Feed
import leets.leenk.domain.feed.domain.repository.FeedRepository
import leets.leenk.domain.feed.test.FeedTestFixture
import leets.leenk.domain.feed.test.UserTestFixture
import leets.leenk.domain.user.domain.entity.UserBlock
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.SliceImpl
import java.time.LocalDateTime
import java.util.Optional

class FeedGetServiceTest :
    BehaviorSpec({
        val feedRepository = mockk<FeedRepository>()
        val feedGetService = FeedGetService(feedRepository)

        Given("삭제되지 않은 피드가 존재할 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)

            every { feedRepository.findByDeletedAtIsNullAndId(1L) } returns Optional.of(feed)

            When("ID로 피드를 조회하면") {
                val result = feedGetService.findById(1L)

                Then("해당 피드가 반환되어야 한다") {
                    result shouldBe feed
                    verify { feedRepository.findByDeletedAtIsNullAndId(1L) }
                }
            }
        }

        Given("삭제되지 않은 피드가 존재하지 않을 때") {
            every { feedRepository.findByDeletedAtIsNullAndId(99L) } returns Optional.empty()

            When("ID로 피드를 조회하면") {
                Then("FeedNotFoundException이 발생해야 한다") {
                    shouldThrow<FeedNotFoundException> {
                        feedGetService.findById(99L)
                    }
                    verify { feedRepository.findByDeletedAtIsNullAndId(99L) }
                }
            }
        }

        Given("차단된 사용자 목록이 있을 때") {
            val blocker = UserTestFixture.createUser(1L, "blocker")
            val blocked1 = UserTestFixture.createUser(2L, "blocked1")
            val blocked2 = UserTestFixture.createUser(3L, "blocked2")
            val blockedUsers =
                listOf(
                    UserBlock
                        .builder()
                        .blocker(blocker)
                        .blocked(blocked1)
                        .build(),
                    UserBlock
                        .builder()
                        .blocker(blocker)
                        .blocked(blocked2)
                        .build(),
                )

            val pageable = PageRequest.of(0, 10)
            val slice = SliceImpl<Feed>(emptyList())

            every { feedRepository.findAllByDeletedAtIsNullWithUser(pageable, listOf(2L, 3L)) } returns slice

            When("전체 피드를 조회하면") {
                val result = feedGetService.findAll(pageable, blockedUsers)

                Then("차단된 사용자를 제외한 피드 목록이 반환되어야 한다") {
                    result shouldBe slice
                    verify { feedRepository.findAllByDeletedAtIsNullWithUser(pageable, listOf(2L, 3L)) }
                }
            }
        }

        Given("특정 사용자의 피드를 조회할 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val pageable = PageRequest.of(0, 5)
            val slice = SliceImpl<Feed>(emptyList())

            every { feedRepository.findAllByUserAndDeletedAtIsNull(user, pageable) } returns slice

            When("사용자의 피드를 조회하면") {
                val result = feedGetService.findAllByUser(user, pageable)

                Then("해당 사용자의 피드 목록이 반환되어야 한다") {
                    result shouldBe slice
                    verify { feedRepository.findAllByUserAndDeletedAtIsNull(user, pageable) }
                }
            }
        }

        Given("이전 피드 네비게이션을 위한 피드들이 있을 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val base = LocalDateTime.of(2025, 12, 22, 16, 0)
            val current = FeedTestFixture.createFeedWithCreateDate(100L, user, base)
            val first = FeedTestFixture.createFeedWithCreateDate(1L, user, base.plusMinutes(1))
            val second = FeedTestFixture.createFeedWithCreateDate(2L, user, base.plusMinutes(2))
            val third = FeedTestFixture.createFeedWithCreateDate(3L, user, base.plusMinutes(3))

            every { feedRepository.findPrevFeeds(base, null, PageRequest.of(0, 3)) } returns
                mutableListOf(first, second, third)

            When("이전 피드 2개를 요청하면") {
                val result = feedGetService.findPrevFeedsWithHasMore(current, emptyList(), 2)

                Then("요청한 개수만큼 반환하고 hasMore가 true여야 한다") {
                    result.hasMore shouldBe true
                    result.feeds shouldContainExactly listOf(second, first)
                    verify { feedRepository.findPrevFeeds(base, null, PageRequest.of(0, 3)) }
                }
            }
        }

        Given("다음 피드 네비게이션을 위한 피드들이 있을 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val blocked = UserTestFixture.createUser(2L, "blocked")
            val block =
                UserBlock
                    .builder()
                    .blocker(user)
                    .blocked(blocked)
                    .build()
            val base = LocalDateTime.of(2025, 12, 22, 16, 0)
            val current = FeedTestFixture.createFeedWithCreateDate(100L, user, base)
            val first = FeedTestFixture.createFeedWithCreateDate(1L, user, base.minusMinutes(1))
            val second = FeedTestFixture.createFeedWithCreateDate(2L, user, base.minusMinutes(2))
            val third = FeedTestFixture.createFeedWithCreateDate(3L, user, base.minusMinutes(3))

            every { feedRepository.findNextFeeds(base, listOf(2L), PageRequest.of(0, 3)) } returns
                mutableListOf(first, second, third)

            When("다음 피드 2개를 요청하면") {
                val result = feedGetService.findNextFeedsWithHasMore(current, listOf(block), 2)

                Then("요청한 개수만큼 반환하고 hasMore가 true여야 한다") {
                    result.hasMore shouldBe true
                    result.feeds shouldContainExactly listOf(first, second)
                    verify { feedRepository.findNextFeeds(base, listOf(2L), PageRequest.of(0, 3)) }
                }
            }
        }
    })
